# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/nsDisplayList.cpp
# Commit: a69c40accbf0
# Full Hash: a69c40accbf099d041417ce4da9b82829f044167
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2023-12-06 21:30:12
# Regressor Bug: 1860328
# File Overlap Count: 2
# Description:
#   Bug 1860328 - Fix a bug where caret misses when dragging and dropping r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D192377
# ==============================================================================

diff -r c9558e7c5c2d -r a69c40accbf0 layout/painting/nsDisplayList.cpp
--- a/layout/painting/nsDisplayList.cpp	Wed Dec 06 16:17:57 2023 +0000
+++ b/layout/painting/nsDisplayList.cpp	Wed Dec 06 16:29:18 2023 +0000
@@ -653,7 +653,6 @@
       mCurrentContainerASR(nullptr),
       mCurrentFrame(aReferenceFrame),
       mCurrentReferenceFrame(aReferenceFrame),
-      mCaretFrame(nullptr),
       mScrollInfoItemsForHoisting(nullptr),
       mFirstClipChainToDestroy(nullptr),
       mTableBackgroundSet(nullptr),
@@ -721,21 +720,6 @@
       mRetainingDisplayList && StaticPrefs::layout_display_list_retain_sc();
 }
 
-static PresShell* GetFocusedPresShell() {
-  nsPIDOMWindowOuter* focusedWnd =
-      nsFocusManager::GetFocusManager()->GetFocusedWindow();
-  if (!focusedWnd) {
-    return nullptr;
-  }
-
-  nsCOMPtr<nsIDocShell> focusedDocShell = focusedWnd->GetDocShell();
-  if (!focusedDocShell) {
-    return nullptr;
-  }
-
-  return focusedDocShell->GetPresShell();
-}
-
 void nsDisplayListBuilder::BeginFrame() {
   nsCSSRendering::BeginFrameTreesLocked();
 
@@ -745,26 +729,6 @@
   mInTransform = false;
   mInFilter = false;
   mSyncDecodeImages = false;
-
-  if (!mBuildCaret) {
-    return;
-  }
-
-  RefPtr<PresShell> presShell = GetFocusedPresShell();
-  if (presShell) {
-    RefPtr<nsCaret> caret = presShell->GetCaret();
-    mCaretFrame = caret->GetPaintGeometry(&mCaretRect);
-
-    // The focused pres shell may not be in the document that we're
-    // painting, or be in a popup. Check if the display root for
-    // the caret matches the display root that we're painting, and
-    // only use it if it matches.
-    if (mCaretFrame &&
-        nsLayoutUtils::GetDisplayRootFrame(mCaretFrame) !=
-            nsLayoutUtils::GetDisplayRootFrame(mReferenceFrame)) {
-      mCaretFrame = nullptr;
-    }
-  }
 }
 
 void nsDisplayListBuilder::AddEffectUpdate(dom::RemoteBrowser* aBrowser,
@@ -804,7 +768,6 @@
   FreeClipChains();
   FreeTemporaryItems();
   nsCSSRendering::EndFrameTreesLocked();
-  mCaretFrame = nullptr;
 }
 
 void nsDisplayListBuilder::MarkFrameForDisplay(nsIFrame* aFrame,
@@ -1139,11 +1102,27 @@
     return;
   }
 
+  RefPtr<nsCaret> caret = state->mPresShell->GetCaret();
+  // This code run for each pres shell and caret->GetPaintGeometry
+  // will return nullptr for invisible caret. So only one caret
+  // can be painted at a time.
+  state->mCaretFrame = caret->GetPaintGeometry(&mCaretRect);
+
+  // Check if the display root for the caret matches the display
+  // root that we're painting, and only use it if it matches. Likely
+  // we only need this for popup.
+  if (state->mCaretFrame &&
+      nsLayoutUtils::GetDisplayRootFrame(state->mCaretFrame) !=
+          nsLayoutUtils::GetDisplayRootFrame(aReferenceFrame)) {
+    state->mCaretFrame = nullptr;
+  }
+
   // Caret frames add visual area to their frame, but we don't update the
   // overflow area. Use flags to make sure we build display items for that frame
   // instead.
-  if (mCaretFrame && mCaretFrame->PresShell() == state->mPresShell) {
-    MarkFrameForDisplay(mCaretFrame, aReferenceFrame);
+  if (state->mCaretFrame) {
+    MOZ_ASSERT(state->mCaretFrame->PresShell() == state->mPresShell);
+    MarkFrameForDisplay(state->mCaretFrame, aReferenceFrame);
   }
 }
 